File: py/ovationincentives_sdk/features.py
# Ovationincentives SDK feature factory

from ovationincentives_sdk.feature.base_feature import OvationincentivesBaseFeature
from ovationincentives_sdk.feature.debug_feature import OvationincentivesDebugFeature
from ovationincentives_sdk.feature.idempotency_feature import OvationincentivesIdempotencyFeature
from ovationincentives_sdk.feature.metrics_feature import OvationincentivesMetricsFeature
from ovationincentives_sdk.feature.paging_feature import OvationincentivesPagingFeature
from ovationincentives_sdk.feature.ratelimit_feature import OvationincentivesRatelimitFeature
from ovationincentives_sdk.feature.retry_feature import OvationincentivesRetryFeature
from ovationincentives_sdk.feature.test_feature import OvationincentivesTestFeature
from ovationincentives_sdk.feature.timeout_feature import OvationincentivesTimeoutFeature


_FEATURES = {
    "base": lambda: OvationincentivesBaseFeature(),
    "debug": lambda: OvationincentivesDebugFeature(),
    "idempotency": lambda: OvationincentivesIdempotencyFeature(),
    "metrics": lambda: OvationincentivesMetricsFeature(),
    "paging": lambda: OvationincentivesPagingFeature(),
    "ratelimit": lambda: OvationincentivesRatelimitFeature(),
    "retry": lambda: OvationincentivesRetryFeature(),
    "test": lambda: OvationincentivesTestFeature(),
    "timeout": lambda: OvationincentivesTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES

File: php/features.php
<?php
declare(strict_types=1);

// Ovationincentives SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class OvationincentivesFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new OvationincentivesBaseFeature();
            case "debug":
                return new OvationincentivesDebugFeature();
            case "idempotency":
                return new OvationincentivesIdempotencyFeature();
            case "metrics":
                return new OvationincentivesMetricsFeature();
            case "paging":
                return new OvationincentivesPagingFeature();
            case "ratelimit":
                return new OvationincentivesRatelimitFeature();
            case "retry":
                return new OvationincentivesRetryFeature();
            case "test":
                return new OvationincentivesTestFeature();
            case "timeout":
                return new OvationincentivesTimeoutFeature();
            default:
                return new OvationincentivesBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}
